dolphin/db/sqlalchemy/api.py

```diff
@@ -295,15 +295,142 @@ def _process_storage_info_filters(query, filters):
     return query
 
 
-def volume_get(context, volume_id):
+def _volume_get_query(context, session=None):
+    return model_query(context, models.Volume, session=session)
+
+
+def _volume_get(context, volume_id, session=None):
+    result = (_volume_get_query(context, session=session)
+              .filter_by(id=volume_id)
+              .first())
+
+    if not result:
+        raise exception.VolumeNotFound(id=volume_id)
+
+    return result
+
+
+def volume_create(context, values):
+    """Create a volume."""
+    if not values.get('id'):
+        values['id'] = uuidutils.generate_uuid()
+
+    vol_ref = models.Volume()
+    vol_ref.update(values)
+
+    session = get_session()
+    with session.begin():
+        session.add(vol_ref)
+
+    return _volume_get(context,
+                       vol_ref['id'],
+                       session=session)
+
+
+def volumes_create(context, volumes):
+    """Create multiple volumes."""
+    session = get_session()
+    vol_refs = []
+    with session.begin():
+
+        for vol in volumes:
+            LOG.debug('adding new volume for original_id {0}:'
+                      .format(vol.get('original_id')))
+            if not vol.get('id'):
+                vol['id'] = uuidutils.generate_uuid()
+
+            vol_ref = models.Volume()
+            vol_ref.update(vol)
+            vol_refs.append(vol_ref)
+
+        session.add_all(vol_refs)
+
+    return vol_refs
+
+
+def volumes_delete(context, volumes_id_list):
+    """Delete multiple volumes."""
+    session = get_session()
+    with session.begin():
+        for vol_id in volumes_id_list:
+            LOG.debug('deleting volume {0}:'.format(vol_id))
+            query = _volume_get_query(context, session)
+            result = query.filter_by(id=vol_id).delete()
+
+            if not result:
+                LOG.error(exception.VolumeNotFound(id=vol_id))
```

Comment:
Do we need this ? If it is exception.VolumeNotFound , it would have raised  in  _volume_get_query itself and the control will never come here!

Reply:
in _volume_get_query(), there is no parameter for passing vol_id. so exception can not be found there.

Reply:
done